src/v/archival/tests/async_data_uploader_test.cc

```diff
@@ -0,0 +1,967 @@
+    // Produce a test case that starts at config batch
+    auto start_fn = [&](size_t ix_size) {
+        auto i = random_generators::get_int((size_t)0, ix_config.size() - 1);
+        auto ix_first = ix_config[i];
+        auto ix_second = random_generators::get_int(ix_first, ix_size - 1);
+        return std::make_pair(ix_first, ix_second);
+    };
+    // Produce a test case that end at config batch
+    auto stop_fn = [&](size_t) {
+        auto i = random_generators::get_int((size_t)0, ix_config.size() - 1);
+        auto ix_second = ix_config[i];
+        auto ix_first = random_generators::get_int((size_t)0, ix_second);
+        return std::make_pair(ix_first, ix_second);
+    };
+    // Produce a test case that starts and stops at config batches
+    auto full_fn = [&](size_t) {
+        auto i = random_generators::get_int((size_t)0, ix_config.size() - 1);
+        auto j = random_generators::get_int(i, ix_config.size() - 1);
+        auto ix_first = ix_config[i];
+        auto ix_second = ix_config[j];
+        return std::make_pair(ix_first, ix_second);
+    };
+    get_random_test_cases_impl(out, map, num_results / 3, start_fn);
+    get_random_test_cases_impl(out, map, num_results / 3, stop_fn);
+    get_random_test_cases_impl(out, map, num_results / 3, full_fn);
+}
+
+FIXTURE_TEST(
+  test_async_segment_upload_random_not_compacted, async_data_uploader_fixture) {
+#ifdef NDEBUG
+    vlog(
+      test_log.info,
+      "Seed used for the test: {}",
+      random_generators::internal::seed);
+    create_topic();
+    for (int i = 0; i < 10; i++) {
+        random_records_generator generator;
+        produce_data(100, generator);
+        ss::abort_source as;
+        get_test_partition()
+          ->archival_meta_stm()
+          ->cleanup_archive(
+            model::offset(0), 0, ss::lowres_clock::now() + 1s, as)
+          .get();
+        roll_segment();
+    }
+    auto log_map = get_log_map();
+    auto partition = get_test_partition();
+
+    std::vector<fuzz_test_case> cases;
+    get_random_test_cases(cases, log_map, 100);
+    get_random_single_batch_test_cases(cases, log_map, 100);
+    get_random_test_cases_that_start_on_config_batches(cases, log_map, 100);
+
+    const auto& offsets = partition->log()->offsets();
+    vlog(
+      test_log.info,
+      "Partition offset range: {}-{}",
+      offsets.start_offset,
+      offsets.committed_offset);
+
+    for (auto tc : cases) {
+        vlog(test_log.info, "Test case, offsets: {}-{}", tc.base, tc.last);
+
+        auto range = inclusive_offset_range(tc.base, tc.last);
+        auto actual = read_offset_range(range);
+        auto expected = load_log_segment_concat(range);
+
+        BOOST_REQUIRE(actual.has_value());
+        BOOST_REQUIRE_EQUAL(actual.value(), expected);
+        BOOST_REQUIRE_EQUAL(actual.value().size_bytes(), tc.expected_size);
+    }
+#endif
+}
+
+FIXTURE_TEST(
+  test_async_segment_upload_random_compacted, async_data_uploader_fixture) {
+#ifdef NDEBUG
```

Comment:
nit: if you switch over to gtest, you can have these tests call `GTEST_SKIP()`, so at least when they're run they're clearly labeled "skipped" instead of opaquely being no-ops

Reply:
switching to gtest won't happen in this PR but I consider doing this in the near future